lib/Parse/ParseExpr.cpp

```diff
@@ -2744,6 +2744,79 @@ void Parser::ParseBlockId(SourceLocation CaretLoc) {
   Actions.ActOnBlockArguments(CaretLoc, DeclaratorInfo, getCurScope());
 }
 
+IdentifierInfo *Parser::getBoundsExpressionStart() {
+  if (Tok.getKind() == tok::identifier) {
+    IdentifierInfo *Ident = Tok.getIdentifierInfo();
+    if (Ident == Ident_bounds || Ident == Ident_byte_count || Ident == Ident_count) {
+      return Ident;
+    }
+  }
+  return nullptr;
+}
+
+ExprResult Parser::ParseBoundsExpression() {
+  IdentifierInfo *Ident = getBoundsExpressionStart();
+  if (!Ident) {
+    Diag(Tok, diag::err_expected_bounds_expr);
+    return ExprError();
+  }
+
+  SourceLocation BoundsKWLoc = Tok.getLocation();
+  ConsumeToken();
+  if (ExpectAndConsume(tok::l_paren)) {
+    return ExprError();
+  }
+
+  ExprResult Result;
+  if (Ident == Ident_byte_count || Ident == Ident_count) {
+    Result = Actions.CorrectDelayedTyposInExpr(ParseAssignmentExpression());
+    CountBoundsExpr::Kind CountKind = Ident == Ident_count ?
+      CountBoundsExpr::Kind::Count : CountBoundsExpr::Kind::Byte_Count;
+    if (!Result.isInvalid()) {
+      Result = Actions.ActOnCountBoundsExpr(BoundsKWLoc, CountKind, Result.get(),
+                                           Tok.getLocation());
+    } else {
+      Result = ExprError();
+    }
+  } else if (Ident == Ident_bounds) {
+    bool FoundNullaryOperator = false;
+
+    if (Tok.getKind() == tok::identifier) {
+      IdentifierInfo *NextIdent = Tok.getIdentifierInfo();
+      if (NextIdent == Ident_none) {
+        FoundNullaryOperator = true;
+        ConsumeToken();
+        Result = Actions.ActOnNullaryBoundsExpr(BoundsKWLoc, NullaryBoundsExpr::Kind::None, Tok.getLocation());
+      }
+    }
+
+    if (!FoundNullaryOperator) {
+      ExprResult LowerBound = Actions.CorrectDelayedTyposInExpr(ParseAssignmentExpression());
+      if (LowerBound.isInvalid()) {
+         Result = ExprError();
+      }
+      if (ExpectAndConsume(tok::comma)) {
+          return ExprError();
+      }
+      ExprResult UpperBound = Actions.CorrectDelayedTyposInExpr(ParseAssignmentExpression());
+      if (!UpperBound.isInvalid()) {
+        Result = Actions.ActOnRangeBoundsExpr(BoundsKWLoc, LowerBound.get(), UpperBound.get(), Tok.getLocation());
+      } else {
+         Result = ExprError();
+      }
+    }
+  } else {
+    Diag(Tok, diag::err_expected_bounds_expr);
+    Result = ExprError();
+  }
+
+  if (ExpectAndConsume(tok::r_paren)) {
```

Comment:
If 'Result' is already an error at this point in the flow of the function, would it be preferable to return it rather than to re-call ExprError()?  (Or perhaps this might only be more valuable in the future if we add more context / explicit messages to the errors?